File: flutter/library/lib/src/tsp_star_rating.dart
import 'package:flutter/material.dart';
import 'star_planet_theme.dart';

/// Interactive star rating.
class TspStarRating extends StatelessWidget {
  const TspStarRating({
    super.key,
    this.value = 0,
    this.max = 5,
    this.disabled = false,
    this.variant = 'default',
    this.theme = StarPlanetTheme.sky,
    this.onChanged,
  });

  final int value;
  final int max;
  final bool disabled;
  final String variant;
  final StarPlanetTheme theme;
  final ValueChanged<int>? onChanged;

  @override
  Widget build(BuildContext context) {
    final count = max.clamp(1, 10);
    final readonly = variant == 'readonly' || disabled;

    return Opacity(
      opacity: disabled ? 0.45 : 1,
      child: Semantics(
        slider: true,
        value: '$value',
        child: Row(
          mainAxisSize: MainAxisSize.min,
          children: [
            for (var i = 1; i <= count; i++)
              GestureDetector(
                onTap: readonly ? null : () => onChanged?.call(i),
                child: Padding(
                  padding: const EdgeInsets.all(2),
                  child: Text(
                    '★',
                    style: TextStyle(
                      fontSize: 22,
                      height: 1,
                      color: i <= value ? theme.brandPrimary : theme.borderDefault,
                    ),
                  ),
                ),
              ),
          ],
        ),
      ),
    );
  }
}

File: react-web/library/src/components/TspStarRating.js
import { React, h, cx, themed, starPlanetTheme } from './_shared.js';

/** Interactive star rating. */
export function TspStarRating({
  value = 0,
  max = 5,
  disabled = false,
  variant = 'default',
  theme = starPlanetTheme,
  onChange,
}) {
  const count = Math.max(1, Math.min(10, Number(max) || 5));
  const readonly = variant === 'readonly' || disabled;
  const stars = [];
  for (let i = 1; i <= count; i += 1) {
    const filled = i <= value;
    stars.push(
      h(
        'button',
        {
          key: i,
          type: 'button',
          className: cx('bc-star-rating__star', filled && 'bc-filled'),
          disabled: readonly,
          'aria-label': `${i}`,
          onClick: readonly ? undefined : () => onChange?.(i),
        },
        '★'
      )
    );
  }
  return h(
    'div',
    {
      className: cx(
        'bc-star-rating',
        `bc-star-rating--${variant}`,
        disabled && 'bc-disabled'
      ),
      style: themed(theme),
      role: 'slider',
      'aria-valuemin': 0,
      'aria-valuemax': count,
      'aria-valuenow': value,
    },
    stars
  );
}
